[CPDNPQ-2142] Freeze financial statements

jebw opened the pull request:
### Context

Ticket: [2142](https://dfedigital.atlassian.net/browse/CPDNPQ-2142)

We have existing functionality in ECF to mark statements as to be paid, this needs bringing across to NPQ

### Changes proposed in this pull request

1. Add controller actions and views to present the authorisation screen with details of the statement
2. Added a component to present the details of the statement on the above screen, currently this uses a stubbed calculator but will switch to use the one being developed as part of CPDNPQ-2141
3. Added a controller action to mark a statement to be paid
4. Added a service to mark declarations as paid
5. Added a service to mark statements as paid
6. Added a background job to mark statements as paid - using the services in (4) and (5)
7. Configured papertrail to track `Statement`, `StatementItem` and `Declaration` models

### Notes

There will be a follow on PR to hook this into the Statement #show screen but thats being developed as part of CPDNPQ-2141 so I've left the integration side of things for now

### Screenshots

![image](https://github.com/user-attachments/assets/11b1a714-cb22-403a-9958-8ebbb35310e6)

### Failed feature specs screenshots

If any of the feature specs would fail, there will be page on the wiki with all
failures:

https://github.com/DFE-Digital/npq-registration/wiki/

jebw commented on config/locales/en.yml:
```diff
@@ -620,6 +629,21 @@ en:
+              title: "Check %{statement_name} statement details before authorising for payment"
+              button: "Authorise for payment"
+              checks_done_legend: Have all necessary assurance checks been done?
+            # tag:
```
Left in because they'll be needed for the follow on PR for integration with CPDNPQ-2141

jebw commented on spec/rails_helper.rb:
```diff
@@ -147,6 +147,12 @@
+  config.before(:each, type: :view) do
+    allow(ActionView::Base)
+      .to receive(:default_form_builder)
+          .and_return GOVUKDesignSystemFormBuilder::FormBuilder
```
Needed to fix view specs defaulting to the regular rails form builder since they are outside the scope of `ApplicationController` - where we are setting the default form builder

jebw commented on app/components/npq_separation/admin/statement_details_component.html.erb:
```diff
@@ -0,0 +1,108 @@
+  </ul>
+  <ul class="booster govuk-list govuk-!-margin-bottom-0">
+    <li>
+      <strong class="govuk-body-s float govuk-!-margin-bottom-0"><%= t("statements.total_retained") %></strong>
```
This would all benefit from a rewrite to use definition lists but for expediency I went with what was already in place on ECF

cwrw commented on app/jobs/statements/mark_as_paid_job.rb:
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+module Statements
+  class MarkAsPaidJob < ApplicationJob
```
could we prevent this from retrying by setting `max_attempts` to 1?

cwrw reviewed (commented):
looks great 🙌  just some very minor comments, happy for this to be merged down to be integrated with the statement calculators

jebw commented:
@cwrw I've updated the PR, I've rebased which confuses the PR but the changes are

1) I've added a change to validate the `marked_as_paid_at` state is present for the Statement state transition to `paid` - this seems like a good idea but if there are concerns about it impacting the data migration this is something we could bring in after separation

https://github.com/DFE-Digital/npq-registration/pull/1894/commits/76ed3693e82ceee2fb9b5397a8f8f217384df158

Other changes

2) Dropped papertrail from `StatementItems` and `Declarations`
3) Added an an exception handler for invalid state transition in the background job - this should prevent DelayedJob retrying but still let us get notified

```diff
diff --git a/app/jobs/statements/mark_as_paid_job.rb b/app/jobs/statements/mark_as_paid_job.rb
index 76e02466..8940c21f 100644
--- a/app/jobs/statements/mark_as_paid_job.rb
+++ b/app/jobs/statements/mark_as_paid_job.rb
@@ -2,6 +2,10 @@
 
 module Statements
   class MarkAsPaidJob < ApplicationJob
+    discard_on StateMachines::InvalidTransition do |_job, exception|
+      Sentry.capture_exception(exception)
+    end
+
     def perform(statement_id:)
       return unless statement_id
 
diff --git a/app/models/declaration.rb b/app/models/declaration.rb
index 4a527519..d22a6237 100644
--- a/app/models/declaration.rb
+++ b/app/models/declaration.rb
@@ -5,9 +5,6 @@ class Declaration < ApplicationRecord
   COURSE_IDENTIFIERS_INELIGIBLE_FOR_UPLIFT = %w[npq-additional-support-offer npq-early-headship-coaching-offer].freeze
   VOIDABLE_STATES = %w[submitted eligible payable ineligible].freeze
 
-  has_paper_trail meta: { note: :version_note }
-  attr_accessor :version_note
-
   belongs_to :application
   belongs_to :cohort
   belongs_to :lead_provider
diff --git a/app/models/statement_item.rb b/app/models/statement_item.rb
index 0ce01932..27310ee9 100644
--- a/app/models/statement_item.rb
+++ b/app/models/statement_item.rb
@@ -2,9 +2,6 @@ class StatementItem < ApplicationRecord
   BILLABLE_STATES = %w[eligible payable paid].freeze
   REFUNDABLE_STATES = %w[awaiting_clawback clawed_back].freeze
 
-  has_paper_trail meta: { note: :version_note }
-  attr_accessor :version_note
-
   belongs_to :statement
   belongs_to :declaration
 
diff --git a/spec/jobs/statements/mark_as_paid_job_spec.rb b/spec/jobs/statements/mark_as_paid_job_spec.rb
index 517276b3..72b1e090 100644
--- a/spec/jobs/statements/mark_as_paid_job_spec.rb
+++ b/spec/jobs/statements/mark_as_paid_job_spec.rb
@@ -2,54 +2,93 @@
 
 require "rails_helper"
 
-RSpec.describe Statements::MarkAsPaidJob do
+RSpec.describe Statements::MarkAsPaidJob, type: :job do
+  include ActiveJob::TestHelper
+
   subject(:job) { described_class.new.perform(statement_id:) }
 
   before do
     allow(Statements::MarkAsPaid).to receive(:new).with(statement).and_return(service)
     allow(service).to receive(:mark).and_return(true)
     allow(Rails.logger).to receive(:warn)
-
-    job
+    allow(Sentry).to receive(:capture_exception).and_return(true)
   end
 
   let(:service) { Statements::MarkAsPaid.new(statement) }
   let(:statement) { create(:statement, :payable) }
   let(:statement_id) { statement.id }
 
-  context "with correct params" do
-    context "when statement is payable" do
-      it "calls the correct service" do
-        expect(service).to have_received(:mark)
+  describe "#perform" do
+    before { job }
+
+    context "with correct params" do
+      context "when statement is payable" do
+        it "calls the correct service" do
+          expect(service).to have_received(:mark)
+        end
+      end
+
+      context "when statement is not payable" do
+        let(:statement) { create(:statement, :paid) }
+
+        it "does not call the service" do
+          expect(Statements::MarkAsPaid).not_to have_received(:new)
+        end
       end
     end
 
-    context "when statement is not payable" do
-      let(:statement) { create(:statement, :paid) }
+    context "with incorrect params" do
+      let(:statement_id) { SecureRandom.uuid }
 
       it "does not call the service" do
         expect(Statements::MarkAsPaid).not_to have_received(:new)
       end
+
+      it "logs a warning" do
+        expect(Rails.logger)
+          .to have_received(:warn)
+              .with("Statement could not be found - statement_id: #{statement_id}")
+      end
+    end
+
+    context "with no params" do
+      let(:statement_id) { nil }
+
+      it { is_expected.to be_nil }
     end
   end
 
-  context "with incorrect params" do
-    let(:statement_id) { SecureRandom.uuid }
+  describe "#perform_later" do
+    subject(:job) { described_class.perform_later(statement_id:) }
 
-    it "does not call the service" do
-      expect(Statements::MarkAsPaid).not_to have_received(:new)
+    it "enqueues the job exactly once" do
+      expect { job }.to have_enqueued_job(described_class).exactly(:once).on_queue("default")
     end
 
-    it "logs a warning" do
-      expect(Rails.logger)
-        .to have_received(:warn)
-            .with("Statement could not be found - statement_id: #{statement_id}")
+    context "with valid job" do
+      before do
+        perform_enqueued_jobs { job }
+      end
+
+      it { expect(Sentry).not_to have_received(:capture_exception) }
+      it { expect(Delayed::Job.count).to be_zero }
     end
-  end
 
-  context "with no params" do
-    let(:statement_id) { nil }
+    context "with invalid job" do
+      before do
+        allow(service).to receive(:mark).and_raise(exception)
 
-    it { is_expected.to be_nil }
+        perform_enqueued_jobs { job }
+      end
+
+      let :exception do
+        StateMachines::InvalidTransition.new(statement,
+                                             Statement.state_machines[:state],
+                                             :mark_paid)
+      end
+
+      it { expect(Sentry).to have_received(:capture_exception) }
+      it { expect(Delayed::Job.count).to be_zero }
+    end
   end
 end
diff --git a/spec/models/declaration_spec.rb b/spec/models/declaration_spec.rb
index 73931ccc..bd4b4a4c 100644
--- a/spec/models/declaration_spec.rb
+++ b/spec/models/declaration_spec.rb
@@ -205,28 +205,6 @@ RSpec.describe Declaration, type: :model do
     end
   end
 
-  describe "paper_trail" do
-    subject { create(:declaration, declaration_type: :started) }
-
-    it "enables paper trail" do
-      expect(subject).to be_versioned
-    end
-
-    it "creates a version with a note" do
-      with_versioning do
-        expect(PaperTrail).to be_enabled
-
-        subject.update!(
-          declaration_type: :completed,
-          version_note: "This is a test",
-        )
-        version = subject.versions.last
-        expect(version.note).to eq("This is a test")
-        expect(version.object_changes["declaration_type"]).to eq(%w[started completed])
-      end
-    end
-  end
-
   describe "#uplift_paid?" do
     let(:declaration_type) { :started }
     let(:targeted_delivery_funding_eligibility) { true }
diff --git a/spec/models/statement_item_spec.rb b/spec/models/statement_item_spec.rb
index e0124d97..7e21340e 100644
--- a/spec/models/statement_item_spec.rb
+++ b/spec/models/statement_item_spec.rb
@@ -77,28 +77,6 @@ RSpec.describe StatementItem, type: :model do
     end
   end
 
-  describe "paper_trail" do
-    subject { create(:statement_item, :eligible) }
-
-    it "enables paper trail" do
-      expect(subject).to be_versioned
-    end
-
-    it "creates a version with a note" do
-      with_versioning do
-        expect(PaperTrail).to be_enabled
-
-        subject.update!(
-          state: :voided,
-          version_note: "This is a test",
-        )
-        version = subject.versions.last
-        expect(version.note).to eq("This is a test")
-        expect(version.object_changes["state"]).to eq(%w[eligible voided])
-      end
-    end
-  end
-
   describe "State transition" do
     describe ".mark_payable" do
       let(:statement_item) { create(:statement_item, :eligible) }
```

cwrw commented on app/jobs/statements/mark_as_paid_job.rb:
```diff
@@ -0,0 +1,21 @@
+
+module Statements
+  class MarkAsPaidJob < ApplicationJob
+    discard_on StateMachines::InvalidTransition do |_job, exception|
```
having a think about this, I've seen it fail in ECF for invalid data that didn't have anything to do with invalid transitions 🤔 I can't remember what it was exactly, but we had to go in and fix the data, I know we said standard error is usually a bad idea but if we're capturing it in Sentry maybe its ok?

jebw replied:
Cool, will change it :+1: then

jebw commented:
@cwrw I've amended the PR - it now rescues StandardError in the Job and I've pulled the validation change out but kept the fixes to the factory data in the specs for a future time when its brought in.

The validation fix itself is now in a separate PR for post separation, see https://github.com/DFE-Digital/npq-registration/pull/1915

I've had to rebase to `main` because of the feature flags in routes change so the actual diff is below

```diff
diff --git a/app/jobs/statements/mark_as_paid_job.rb b/app/jobs/statements/mark_as_paid_job.rb
index 8940c21f..83089870 100644
--- a/app/jobs/statements/mark_as_paid_job.rb
+++ b/app/jobs/statements/mark_as_paid_job.rb
@@ -2,7 +2,7 @@
 
 module Statements
   class MarkAsPaidJob < ApplicationJob
-    discard_on StateMachines::InvalidTransition do |_job, exception|
+    discard_on StandardError do |_job, exception|
       Sentry.capture_exception(exception)
     end
 
diff --git a/app/models/statement.rb b/app/models/statement.rb
index 2d916e21..c3f289d4 100644
--- a/app/models/statement.rb
+++ b/app/models/statement.rb
@@ -40,10 +40,7 @@ class Statement < ApplicationRecord
   state_machine :state, initial: :open do
     state :open
     state :payable
-
-    state :paid do
-      validates :marked_as_paid_at, presence: true
-    end
+    state :paid
 
     event :mark_payable do
       transition [:open] => :payable
diff --git a/spec/models/statement_spec.rb b/spec/models/statement_spec.rb
index 062e7aeb..a462c8f7 100644
--- a/spec/models/statement_spec.rb
+++ b/spec/models/statement_spec.rb
@@ -130,24 +130,12 @@ RSpec.describe Statement, type: :model do
     end
 
     context "when from payable to paid" do
-      let(:statement) { create(:statement, :payable) }
+      let(:statement) { create(:statement, :payable, marked_as_paid_at: 1.week.ago) }
 
-      context "with mark_as_paid_at set" do
-        before { statement.update(marked_as_paid_at: Time.zone.now) }
-
-        it "transitions state to payable" do
-          expect(statement).to be_payable
-          statement.mark_paid!
-          expect(statement).to be_paid
-        end
-      end
-
-      context "without mark_as_paid_at being set" do
-        it "raises an error" do
-          expect(statement.reload).to be_payable
-          expect { statement.mark_paid! }.to raise_error(StateMachines::InvalidTransition)
-          expect(statement.reload).to be_payable
-        end
+      it "transitions state to payable" do
+        expect(statement).to be_payable
+        statement.mark_paid!
+        expect(statement).to be_paid
       end
     end
 
diff --git a/spec/requests/npq_separation/admin/finance/statements/payment_authorisations_controller_spec.rb b/spec/requests/npq_separation/admin/finance/statements/payment_authorisations_controller_spec.rb
index 8fd173c8..f33218a9 100644
--- a/spec/requests/npq_separation/admin/finance/statements/payment_authorisations_controller_spec.rb
+++ b/spec/requests/npq_separation/admin/finance/statements/payment_authorisations_controller_spec.rb
@@ -1,6 +1,6 @@
 require "rails_helper"
 
-RSpec.describe NpqSeparation::Admin::Finance::Statements::PaymentAuthorisationsController, type: :request do
+RSpec.describe NpqSeparation::Admin::Finance::Statements::PaymentAuthorisationsController, :ecf_api_disabled, type: :request do
   include Helpers::NPQSeparationAdminLogin
 
   subject { response }
```

cwrw reviewed (approved):
Thank you @jebw 🙌 looks great!
